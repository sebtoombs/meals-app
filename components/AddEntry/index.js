import { useEffect, useState } from "react";
import {
  Modal,
  ModalOverlay,
  ModalContent,
  ModalHeader,
  ModalFooter,
  ModalBody,
  ModalCloseButton,
  Button,
  VStack,
  Flex,
  Box,
  Skeleton,
  Text,
} from "@chakra-ui/react";
import { DayPicker } from "../DayPicker";
import { useForm, Controller } from "react-hook-form";
import { ErrorMessage } from "@hookform/error-message";
import useMeals from "../../lib/queries/useMeals";
import { yupResolver } from "@hookform/resolvers/yup";
import * as yup from "yup";
import { MdAdd } from "react-icons/md";
import useAddCalendarEntry from "../../lib/mutations/useAddCalendarEntry";
import { format, parse } from "date-fns";

export function AddEntry({ addEntryView }) {
  const [meal, setMeal] = useState();
  const [date, setDate] = useState(addEntryView.state || null);

  useEffect(() => {
    setDate(addEntryView.state);
  }, [addEntryView.state]);

  const addCalendarEntry = useAddCalendarEntry();

  const onSubmit = async (e) => {
    e.preventDefault();

    try {
      await addCalendarEntry.mutateAsync({
        mealId: meal.id,
        date,
      });

      addEntryView.onClose();
    } catch (error) {
      console.error(error);
      // TODO handle error
    }
  };

  useEffect(() => {
    if (!addEntryView.isOpen && meal) setMeal(null);
    if (!addEntryView.isOpen && date) setDate(null);
  }, [addEntryView.isOpen]);

  return (
    <AddEntryView
      {...addEntryView}
      {...{ meal, setMeal, date, setDate }}
      onSubmit={onSubmit}
    />
  );
}

export function AddEntryView({
  isOpen,
  onOpen,
  onClose,
  onSubmit,
  meal,
  setMeal,
  date,
  setDate,
}) {
  return (
    <Modal isOpen={isOpen} onClose={onClose} motionPreset="slideInRight">
      <ModalOverlay />
      <ModalContent as="form" onSubmit={onSubmit} bg="gray.100">
        <ModalHeader>Add Meal</ModalHeader>
        <ModalCloseButton />
        <ModalBody>
          {!meal && <AddEntryMealList {...{ meal, setMeal }} />}
          {!!meal && (
            <DayPicker
              name={"date"}
              value={date}
              onChange={(dateString) => {
                setDate(parse(dateString, "yyyy-MM-dd", new Date()));
              }}
            />
          )}
        </ModalBody>

        <ModalFooter>
          <Button mr={3} onClick={onClose}>
            Close
          </Button>
          {!!meal && (
            <Button
              colorScheme="teal"
              type="submit"
              isDisabled={!meal || !date}
            >
              Add to Calendar
            </Button>
          )}
        </ModalFooter>
      </ModalContent>
    </Modal>
  );
}

export function AddEntryMealList({ meal, setMeal }) {
  // Filter/search state will go here

  const mealOptions = useMeals();

  if (
    mealOptions.status === "loading" ||
    (mealOptions.status === "success" && mealOptions.data.length)
  )
    return (
      <AddEntryMealListView mealOptions={mealOptions} {...{ meal, setMeal }} />
    );
  else if (mealOptions.status === "success" && !mealOptions.data.length)
    return <AddEntryMealListEmptyView />;

  return <AddEntryMealListErrorView meals={mealOptions} />;
}

export function AddEntryMealListView({ mealOptions, meal, setMeal }) {
  return (
    <VStack spacing="4" sx={{ "&>*": { minWidth: "100%" } }}>
      {mealOptions?.data?.map((meal, index) => (
        <Box
          as="button"
          overflow="hidden"
          rounded="md"
          bg="white"
          key={index}
          textAlign="left"
          _focus={{
            shadow: "outline",
            outline: "2px solid transparent",
            outlineOffset: "2px",
          }}
          onClick={() => (meal.id ? setMeal(meal) : null)}
          disabled={!meal.id}
        >
          <Flex>
            {/* TODO image goes here */}
            <Box bg="gray.300" w="12" />

            {meal.id ? (
              <Flex
                flexGrow="1"
                p="2"
                flexDir="column"
                justifyContent="center"
                fontWeight="medium"
              >
                {meal.name}
              </Flex>
            ) : (
              <Box flexGrow="1">
                <Skeleton width="100%" height="100%" />
              </Box>
            )}
            <Flex p="2" flexDir="column" justifyContent="center">
              <Box
                as="span"
                rounded="md"
                bg="teal.500"
                color="white"
                w="8"
                h="8"
                display="flex"
                alignItems="center"
                justifyContent="center"
                fontSize="2xl"
              >
                <MdAdd />
              </Box>
            </Flex>
          </Flex>
        </Box>
      ))}
    </VStack>
  );
}

export function AddEntryMealListEmptyView() {
  return (
    <Flex justifyContent="center">
      <Text>No meals to select.</Text>
    </Flex>
  );
}

export function AddEntryMealListErrorView({ meals }) {
  return (
    <Flex justifyContent="center">
      <Text>Something went wrong.</Text>
    </Flex>
  );
}
