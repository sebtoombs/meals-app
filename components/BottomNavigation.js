import React, { useState } from "react";
import styled from "styled-components/macro";
import tw from "tailwind.macro";
import { MdRestaurantMenu, MdList } from "react-icons/md";
import { GiMeal } from "react-icons/gi";
import { useStateValue } from "../lib/state";
import Link from "next/link";
import { withRouter } from "next/router";

const ItemWrapper = styled.span`
  ${tw`block inline-flex flex-column items-center justify-center flex-wrap text-gray-600`}
  svg {
    ${tw`block min-w-full text-2xl`}
  }
  span {
    ${tw`block min-w-full text-sm text-center`}
  }
`;
const NavigationItem = styled.a`
  ${tw`inline-flex cursor-pointer relative items-center justify-center px-4 py-2 border-b-2 border-transparent`} flex: 1;
  min-width: 80px;
  max-width: 168px;

${props => (props.active ? tw`bg-teal-100 border-teal-500` : null)}
  & ${ItemWrapper} {
    ${props => (props.active ? tw`text-teal-500` : null)}
  }
`;

const BottomNavigationStyled = styled.div`
  ${tw`fixed bg-gray-100 w-full flex`} bottom: 0;
  max-width: 28rem;
`;

const getCurrentPath = pathname => {
  pathname = pathname
    .split("/")
    .filter(part => !!part)
    .slice(0, 1);

  return pathname.length ? pathname[0] : "index";
};

const BottomNavigation = props => {
  const { router } = props;

  const path = getCurrentPath(router.pathname);

  //const [{ activePage }, dispatch] = useStateValue();

  /*const setActivePage = page => {
    return () => {
      dispatch({ type: "setActivePage", page });
    };
  };*/

  return (
    <BottomNavigationStyled>
      <Link href={`/`}>
        <NavigationItem active={path === "index"}>
          <ItemWrapper>
            <MdRestaurantMenu />
            <span>Menu</span>
          </ItemWrapper>
        </NavigationItem>
      </Link>
      <Link href={`/meals`}>
        <NavigationItem active={path === "meals"}>
          <ItemWrapper>
            <GiMeal />
            <span>Meals</span>
          </ItemWrapper>
        </NavigationItem>
      </Link>
      <NavigationItem active={path === "shopping"}>
        <ItemWrapper>
          <MdList />
          <span>Shopping</span>
        </ItemWrapper>
      </NavigationItem>
    </BottomNavigationStyled>
  );
};
export default withRouter(BottomNavigation);
