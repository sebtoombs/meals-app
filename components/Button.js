import styled from "styled-components/macro";
import tw from "tailwind.macro";

const ButtonStyled = styled.button`
${tw`bg-gray-400 text-black px-3 py-1 text-lg rounded-lg inline-flex cursor-pointer items-center text-center border border-transparent justify-center`}
  transition: 0.2s background-color ease, 0.2s color ease, 0.2s color ease;
  ${props => (props.shadow ? tw`shadow-xl` : null)}

  ${props => (props.size && props.size === "sm" ? tw`px-2 text-base` : null)}
  ${props => (props.size && props.size === "lg" ? tw`px-5 py-3 text-xl` : null)}
  ${props =>
    props.size && props.size === "xl" ? tw`px-8 py-3 text-2xl` : null}
  ${props => (props.disabled ? tw`bg-gray-200 text-gray-600` : null)}
  ${props => (props.block ? tw`flex w-full` : null)}

${props => (props.danger ? tw`bg-red-500 text-white` : null)}
  ${props => (props.primary ? tw`bg-teal-500 text-white` : null)}
  ${props => (props.link ? tw`bg-transparent text-gray-700 underline` : null)}

${props =>
  props.outline ? tw`border-gray-700 bg-transparent text-gray-700` : null}
  ${props =>
    props.danger && props.outline
      ? tw`text-red-500 bg-transparent border-red-500`
      : null}

  ${props =>
    props.primary && props.outline
      ? tw`text-teal-500 bg-transparent border-teal-500`
      : null}

  ${props => (props.disabled ? tw`opacity-50 cursor-default` : null)}

  svg {
    ${tw`text-2xl`}
  }
`;

const Button = props => {
  let passProps = { ...props };

  const propList = ["primary", "block", "link"];

  propList.map(propName => {
    if (typeof props[propName] !== "undefined") {
      passProps[propName] = "true";
    }
  });

  if (props.icon) {
    passProps.children = (
      <>
        <span css={tw`mr-3`}>{props.icon}</span>{" "}
        <span css={tw`mr-2`}>{props.children}</span>
      </>
    );
  }
  if (props.iconAfter) {
    passProps.children = (
      <>
        <span css={tw`ml-2`}>{props.children}</span>{" "}
        <span css={tw`ml-3`}>{props.iconAfter}</span>
      </>
    );
  }

  return (
    <ButtonStyled as={passProps.tag} {...passProps}>
      {passProps.children}
    </ButtonStyled>
  );
};
export default Button;
