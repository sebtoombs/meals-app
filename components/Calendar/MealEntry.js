import { Box, AspectRatio, Text } from "@chakra-ui/react";

export default function MealEntry({ entry, ...props }) {
  return (
    <Box
      rounded="md"
      overflow="hidden"
      as="button"
      cursor="pointer"
      textAlign="left"
      w="100%"
      _focus={{
        boxShadow: "outline",
        outlineOffset: "2px",
        outline: "2px transparent",
      }}
      bg="white"
      {...props}
    >
      {/* TODO image goes here */}
      <AspectRatio ratio={16 / 7}>
        <Box bg="gray.100" />
      </AspectRatio>
      <Text fontWeight="medium" p="4">
        {entry?.meal?.name}
      </Text>
    </Box>
  );
}
