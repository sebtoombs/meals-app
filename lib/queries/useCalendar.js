import { request, gql } from "graphql-request";
import { useQuery } from "react-query";
import { startOfDay, addDays } from "date-fns";

const endpoint = "http://localhost:3000/api";

export default function useCalendar({
  includeEmpty = false,
  placeholderData = null,
} = {}) {
  return useQuery(
    ["calendar", { includeEmpty }],
    async ({ queryKey }) => {
      const [_key, { includeEmpty }] = queryKey;
      const { calendar } = await request(
        endpoint,
        gql`
          query CalendarQuery($includeEmpty: Boolean) {
            calendar(includeEmpty: $includeEmpty) {
              date
              entries {
                id
                date
                meal {
                  id
                  name
                }
              }
            }
          }
        `,
        { includeEmpty }
      );

      return calendar;
    },
    {
      placeholderData,
    }
  );
}
