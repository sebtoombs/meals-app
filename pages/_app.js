import { ChakraProvider } from "@chakra-ui/react";
import { QueryClient, QueryClientProvider } from "react-query";
import { ReactQueryDevtools } from "react-query/devtools";

const queryClient = new QueryClient();

// Abstract the wrapper so we can write tests
export function Wrapper({ children }) {
  return (
    <ChakraProvider>
      <QueryClientProvider client={queryClient}>
        {children}
        {process.env.NODE_ENV === "development" && <ReactQueryDevtools />}
      </QueryClientProvider>
    </ChakraProvider>
  );
}

function MyApp({ Component, pageProps }) {
  return (
    <Wrapper>
      <Component {...pageProps} />
    </Wrapper>
  );
}
export default MyApp;
