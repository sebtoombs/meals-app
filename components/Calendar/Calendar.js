import { useState } from "react";
import useCalendar from "../../lib/queries/useCalendar";
import { Global } from "@emotion/react";
import {
  Container,
  Heading,
  VStack,
  Text,
  Button,
  useDisclosure,
  Drawer,
  DrawerOverlay,
  DrawerContent,
  DrawerHeader,
  DrawerBody,
  DrawerCloseButton,
  Wrap,
  WrapItem,
  Box,
  FormControl,
  FormErrorMessage,
  FormLabel,
  forwardRef,
  HStack,
  Skeleton,
  Flex,
  Portal,
  Icon,
  Grid,
  AspectRatio,
} from "@chakra-ui/react";
import { MdAdd } from "react-icons/md";
import { AddEntry } from "../AddEntry";
import CalendarCard, {
  CalendarCardTitle,
  CalendarCardHeader,
  CalendarCardBody,
} from "../CalendarCard";
import { addDays, isSameDay, startOfDay, format, parseISO } from "date-fns";
import { ViewMeal } from "../ViewMeal";
import MealEntry from "./MealEntry";

function useDisclosureWithState() {
  const { isOpen, onOpen, onClose } = useDisclosure();
  const [state, setState] = useState();
  return {
    isOpen,
    state,
    onOpen: (newState) => {
      if (
        newState &&
        ["String", "Boolean", "Array", "Number", "Object"].indexOf(
          newState.constructor.name
        ) === -1
      ) {
        newState = undefined;
      }
      setState(newState);
      onOpen();
    },
    onClose: () => {
      setState();
      onClose();
    },
  };
}

export function Calendar() {
  // Access the client
  //const queryClient = useQueryClient();

  // Queries
  const calendar = useCalendar({ includeEmpty: true });

  // Add entry
  const addEntryView = useDisclosureWithState(); // {isOpen, onOpen, onClose}

  // View meal
  const viewMealView = useDisclosureWithState();

  return (
    <CalendarView
      calendar={calendar}
      addEntryView={addEntryView}
      viewMealView={viewMealView}
    />
  );
}

export function CalendarView({ calendar, addEntryView, viewMealView }) {
  const renderView = () => {
    if (
      calendar.status === "loading" ||
      (calendar.status === "success" && calendar?.data?.length)
    )
      return CalendarListView({
        dayEntries: calendar.data,
        addEntryView,
        status: calendar.status,
        viewMealView,
      });
    else if (calendar.status === "success" && !calendar?.data?.length)
      return CalendarEmptyView({ addEntryView });

    return <CalendarErrorView />;
  };

  return (
    <>
      {renderView()}
      <AddEntry addEntryView={addEntryView} />
      <ViewMeal viewMealView={viewMealView} />
    </>
  );
}

export function CalendarErrorView() {
  return <p>Oh no I haven't built this view!</p>;
}

export function CalendarListView({
  dayEntries = [],
  status = "loading",
  addEntryView,
  viewMealView,
}) {
  const todayDate = startOfDay(new Date());
  const tomorrowDate = addDays(todayDate, 1);
  const entryDateToTitle = (dateISOString, index) => {
    const date = parseISO(dateISOString);
    if (isSameDay(date, todayDate)) return "Today";
    if (isSameDay(date, tomorrowDate)) return "Tomorrow";
    const dayOfWeek = format(date, "EEEE");
    if (index > 6) return `Next ${dayOfWeek}`;
    return dayOfWeek;
  };

  return (
    <>
      <Global
        styles={(theme) => ({
          body: {
            background: theme.colors.gray["50"],
          },
        })}
      />
      <Box p="4">
        <VStack spacing="4">
          {status === "loading" ? (
            <Skeleton width="100%" height="40px" />
          ) : (
            dayEntries?.map((dayEntry, index) => (
              <CalendarCard
                key={index}
                bg={index === 0 ? "blue.100" : "white"}
                border="1px solid"
                borderColor={index === 0 ? "blue.300" : "gray.100"}
              >
                <CalendarCardHeader>
                  <CalendarCardTitle>
                    {entryDateToTitle(dayEntry.date, index)}
                  </CalendarCardTitle>
                </CalendarCardHeader>

                {dayEntry.entries.length ? (
                  dayEntry.entries.length === 1 ? (
                    <Box px="4" pb="4">
                      <MealEntry
                        entry={dayEntry.entries[0]}
                        onClick={() => {
                          viewMealView.onOpen({
                            mealId: dayEntry.entries[0].meal.id,
                            id: dayEntry.entries[0].id,
                          });
                        }}
                      />
                    </Box>
                  ) : (
                    <Grid
                      sx={{
                        "--gutter": "var(--chakra-sizes-4)",
                        scrollSnapType: "x proximity",
                        "&:before,&:after": {
                          content: '""',
                          width: "4",
                        },
                        gridGap: "calc(var(--gutter) / 2)",
                        gridTemplateColumns: "10px",
                        gridTemplateRows: "minmax(100%, 1fr)",
                        gridAutoFlow: "column",
                        gridAutoColumns: `calc(${
                          dayEntry.entries.length === 2 ? "75%" : "50%"
                        } - var(--gutter) * 2)`,
                        overflowX: "scroll",
                        paddingBottom: "calc(.75 * var(--gutter))",
                        marginBottom: "calc(-.25 * var(--gutter))",
                        scrollbarWidth: "none",
                        marginBottom: 0,
                        pt: "4px",
                        pb: "4",
                        "&::-webkit-scrollbar": {
                          display: "none",
                        },
                      }}
                    >
                      {dayEntry.entries.map((entry, index) => (
                        <Box key={index}>
                          <MealEntry
                            entry={entry}
                            bg={
                              dayEntry.entries.length > 1 ? "gray.50" : "white"
                            }
                            onClick={() => {
                              viewMealView.onOpen({
                                mealId: entry.meal.id,
                                id: entry.id,
                              });
                            }}
                          />
                        </Box>
                      ))}
                    </Grid>
                  )
                ) : (
                  <Flex p="4" justifyContent="center">
                    <Button
                      colorScheme="teal"
                      size="sm"
                      leftIcon={<MdAdd />}
                      onClick={() => addEntryView.onOpen(dayEntry.date)}
                    >
                      Add
                    </Button>
                  </Flex>
                )}
              </CalendarCard>
            ))
          )}
        </VStack>
      </Box>
      <Portal>
        <Box position="fixed" right="2" bottom="20">
          <Button
            rounded="full"
            colorScheme="teal"
            display="flex"
            w="14"
            h="14"
            shadow="lg"
            alignItems="center"
            justifyContent="center"
            onClick={addEntryView.onOpen}
          >
            <Box as="span">
              <Icon as={MdAdd} fontSize="2xl" />
              <Text fontSize="xs">Add</Text>
            </Box>
          </Button>
        </Box>
      </Portal>
    </>
  );
}

export function CalendarEmptyView({ addEntryView }) {
  return (
    <Container>
      <VStack py={[10, null, 20]} spacing={8}>
        <Heading>Meals</Heading>
        <Text>There's nothing on your menu yet</Text>
        <Button
          colorScheme="teal"
          leftIcon={<MdAdd />}
          onClick={addEntryView.onOpen}
        >
          Add a meal
        </Button>
      </VStack>
    </Container>
  );
}
