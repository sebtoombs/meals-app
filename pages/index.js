import Head from "next/head";

import { Calendar } from "../components/Calendar";

export default function Home() {
  return (
    <div>
      <Head>
        <title>Meals</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <main>
        <Calendar />
      </main>
    </div>
  );
}
