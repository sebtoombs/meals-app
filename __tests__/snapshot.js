import * as React from "react";
import { Wrapper } from "../pages/_app";
import { Calendar, CalendarView } from "../components/Calendar";
import useCalendar from "../lib/queries/useCalendar";
import { renderHook, act } from "@testing-library/react-hooks";
import { QueryClientProvider, QueryClient, useQuery } from "react-query";
import { request, gql } from "graphql-request";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

// const queryClient = new QueryClient();
// const wrapper = ({ children }) => (
//   <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
// );

const flushPromises = () => new Promise(setImmediate);

test("Calendar renders", async () => {
  const { asFragment } = render(
    <Wrapper>
      <Calendar />
    </Wrapper>
  );
  await flushPromises();
  // expect(await screen.findByText(/pizza/i)).toBeInTheDocument();
  const firstRender = asFragment();
  expect(firstRender).toMatchSnapshot();
});
