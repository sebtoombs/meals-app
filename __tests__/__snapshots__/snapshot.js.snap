// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar renders 1`] = `
<DocumentFragment>
  <div
    class="chakra-container css-1ajyfvc"
  >
    <div
      class="chakra-stack css-1iv4t3s"
    >
      <h2
        class="chakra-heading css-eo8fjx"
      >
        Meals
      </h2>
      <p
        class="chakra-text css-1u8qly9"
      >
        There's nothing on your menu yet
      </p>
      <button
        class="chakra-button css-1u0s644"
        type="button"
      >
        <span
          class="chakra-button__icon css-ws6tz2"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        Add a meal
      </button>
    </div>
  </div>
</DocumentFragment>
`;
